### PR TITLE
Activate Dencun/Cancun upgrade

### LINKED ISSUE #1299: Support Dencun
Dencun upgrade introduced blobs mainly, however, there are small changes that are very interesting for us.

Primarily the changes to `selfdestruct`.

We should also talk about what would it mean for us to support blobs as well. I would like to handle that under discussions.

EDIT: Discussion is here https://github.com/chainwayxyz/citrea/discussions/1300

### PR DESCRIPTION
**Pull request by eyusufatik**
# Description
Adds dencun support to our evm. however, we won't let users submit eip4844 transactions at the mempool level, since there is no DA guarantees of blobs on a centralized sequencer. However, we want to keep compatability so all blob stuff are implemented. The main benefit we get is new opcodes and `seldestruct` change

TODOs:

- [x] Block blob transaction from the rpc
- [x] ~Pass on blob transactions inside evm~ instead we'll trust the sequencer to not include eip4844 txs.
- [ ] Benchmark kzg precompile and decide on enabling it
- [x] ~Make sure blob gas fee settings = 0 does not break anything~ Instead we'll correctly calculate eip4844 fee rates. Since we don't accept any eip4844 txs, this won't have any effect on anything.
- [ ] Activate ef tests for cancun -- though they might be problematic, instead test opcodes individually 
- [x] Write a test where we activate Fork1 hence cancun
- [ ] Update `selfdestruct` in `db_commit.rs`

## Linked Issues
Closes #1299

### CODE REVIEW

File: crates/evm/src/hooks.rs [hunk 1]
```diff
@@ -199,11 +213,11 @@ where
                 .iter()
                 .fold(Bloom::ZERO, |bloom, r| bloom | r.bloom),
             difficulty: U256::ZERO,
-            gas_limit: self.block_env.gas_limit,
+            gas_limit: self.block_env.gas_limit.saturating_to(),
             gas_used,
-            mix_hash: self.block_env.prevrandao,
+            mix_hash: self.block_env.prevrandao.unwrap_or_else(|| B256::default()),
```
Comment (kpp) [inline crates/evm/src/hooks.rs]: unwrap_or_default

File: crates/sovereign-sdk/rollup-interface/src/fork/manager.rs [hunk 2]
```diff
@@ -15,6 +15,7 @@ impl ForkManager {
         // Make sure the list of specs is sorted by the block number at which they activate.
         forks.sort_by_key(|fork| fork.activation_height);
 
+        // TODO: binary search here and assume forks is always sorted
```
Comment (rakanalh) [inline crates/sovereign-sdk/rollup-interface/src/fork/manager.rs:18]: It is sorted above

Comment (eyusufatik) [reply crates/sovereign-sdk/rollup-interface/src/fork/manager.rs:18]: no I meant trust input is sorted no need to sort

File: crates/evm/src/query.rs [hunk 3]
```diff
@@ -710,14 +752,30 @@ impl<C: sov_modules_api::Context> Evm<C> {
...
+            let citrea_spec_id =
+                fork_from_block_number(FORKS.to_vec(), block_env.number.saturating_to()).spec_id;
+            let evm_spec_id = citrea_spec_id_to_evm_spec_id(citrea_spec_id);
+
+            let cfg_env = get_cfg_env(cfg, evm_spec_id);
+            block_env.blob_excess_gas_and_price = if evm_spec_id >= SpecId::CANCUN {
+                Some(BlobExcessGasAndPrice {
+                    excess_blob_gas: 0,
+                    blob_gasprice: 0,
+                })
```
Comment (eyusufatik) [inline crates/evm/src/query.rs]: can use values from header

File: crates/evm/src/evm/conversions.rs [hunk 4]
```diff
@@ -132,3 +128,23 @@ impl From<TransactionSignedAndRecovered> for TransactionSignedEcRecovered {
...
+        coinbase: sealed_header.beneficiary,
+        timestamp: U256::from(sealed_header.timestamp),
+        prevrandao: Some(sealed_header.mix_hash),
+        basefee: U256::from(sealed_header.base_fee_per_gas.unwrap_or_default()),
+        gas_limit: U256::from(sealed_header.gas_limit),
+        difficulty: U256::from(0),
+        blob_excess_gas_and_price: if spec_id >= SpecId::CANCUN {
+            Some(BlobExcessGasAndPrice {
+                excess_blob_gas: 0,
+                blob_gasprice: 0,
```
Comment (eyusufatik) [inline crates/evm/src/evm/conversions.rs]: again

File: crates/evm/src/evm/executor.rs [hunk 5]
```diff
@@ -104,9 +105,23 @@ pub(crate) fn execute_multiple_tx<
...
+                "Blob gas used exceeds block gas limit {:?}",
+                block_gas_limit
+            ))));
+            continue;
+        }
+
         let result_and_state = match evm.transact(tx) {
             Ok(result_and_state) => result_and_state,
             Err(e) => {
+                println!("Error: {:?}", e);
```
Comment (eyusufatik) [inline crates/evm/src/evm/executor.rs]: remove

File: crates/evm/src/tests/genesis_tests.rs [hunk 6]
```diff
@@ -195,7 +166,7 @@ fn genesis_block() {
 fn genesis_head() {
     let (evm, mut working_set) = get_evm(&get_evm_test_config());
     let head = evm.head.get(&mut working_set).unwrap();
-
+    println!("{:?}", head);
```
Comment (eyusufatik) [inline crates/evm/src/tests/genesis_tests.rs]: remove

File: crates/evm/src/tests/call_tests.rs [hunk 7]
```diff
@@ -1642,3 +1643,72 @@ fn test_call_with_block_overrides() {
...
+    }
+    evm.end_soft_confirmation_hook(&soft_confirmation_info, &mut working_set);
+    evm.finalize_hook(&[99u8; 32].into(), &mut working_set.accessory_state());
+
+    let last_block = evm
+        .get_block_by_number(Some(BlockNumberOrTag::Latest), None, &mut working_set)
+        .unwrap()
+        .unwrap();
+
+    println!("{:?}", last_block);
```
Comment (eyusufatik) [inline crates/evm/src/tests/call_tests.rs]: remove and add todo explaining this isn't doing anything :)

**Comment by eyusufatik** [comment]
since some prs now depend on this, I've opened issues for stuff that I wanted to handle in this pr.

File: crates/evm/src/tests/call_tests.rs [hunk 8]
```diff
@@ -1642,3 +1643,71 @@ fn test_call_with_block_overrides() {
...
+        da_slot_txs_commitment: [42u8; 32],
+        pre_state_root: [10u8; 32].to_vec(),
+        current_spec: SovSpecId::Fork1, // wont be Fork1 at height 2 currently but we can trick the spec id
+        pub_key: vec![],
+        deposit_data: vec![],
+        l1_fee_rate,
+        timestamp: 0,
+    };
+
+    // Deploy block hashes contract
```
Comment (rakanalh) [inline crates/evm/src/tests/call_tests.rs]: Irrelevant comment?

File: crates/evm/src/tests/call_tests.rs [hunk 9]
```diff
@@ -1642,3 +1643,71 @@ fn test_call_with_block_overrides() {
     let expected_hash = Bytes::from_iter([2; 32]);
     assert_eq!(call_result, expected_hash);
 }
+
+// TODO: test is not doing anything significant at the moment
```
Comment (rakanalh) [inline crates/evm/src/tests/call_tests.rs:1647]: Can you describe whats left to be done for this test?

**Review by rakanalh — APPROVED** [review]
Approving the PR to have this merged, we can deal with my small comments later if we decide to.

**Pull request merged**